core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLRecordReader.java

```diff
@@ -0,0 +1,305 @@
+    String lastNode = nodes[lastNodeIndex];
+    StringBuilder xmlRecord = new StringBuilder();
+
+    //flag to know if xml record matching to node path has been read and ready to use.
+    boolean xmlRecordReady = false;
+    //flag to know if matching node found as per node path
+    boolean nodeFound = false;
+
+    try {
+      while (reader.hasNext()) {
+        int event = reader.next();
+        switch (event) {
+          case XMLStreamConstants.START_ELEMENT:
+            String nodeNameStart = reader.getLocalName();
+            boolean validHierarchy = true;
+            currentNodeLevelMap.put(nodeLevel, nodeNameStart);
+
+            //check if node hierarchy matches with expected one.
+            for (int j = nodeLevel; j >= 0; j--) {
+              if (!currentNodeLevelMap.get(j).equals(actualNodeLevelMap.get(j))) {
+                validHierarchy = false;
+              }
+            }
+            //check if node hierarchy is valid and it matches with last node in node path
+            //then start appending tag information to create valid XML Record.
+            if (validHierarchy && nodeNameStart.equals(lastNode)) {
+              appendStartTagInformation(nodeNameStart, xmlRecord);
+              //Set flag for valid node path found
+              nodeFound = true;
+              //set file offset
+              currentKey.set(reader.getLocation().getLineNumber());
+            } else if (nodeFound) {
+              //append all child nodes inside valid node path
+              appendStartTagInformation(nodeNameStart, xmlRecord);
+            }
+            nodeLevel++;
+            break;
+          case XMLStreamConstants.CHARACTERS:
+            if (nodeFound) {
+              xmlRecord.append(reader.getText());
+            }
+            break;
+          case XMLStreamConstants.END_ELEMENT:
+            String nodeNameEnd = reader.getLocalName();
+            if (nodeFound) {
+              //add closing tag
+              xmlRecord.append(OPENING_END_TAG_DELIMITER).append(nodeNameEnd).append(CLOSING_END_TAG_DELIMITER);
+              if (nodeNameEnd.equals(lastNode)) {
+                nodeFound = false;
+                //set flag for XML record is ready to emit.
+                xmlRecordReady = true;
+              }
+            }
+            nodeLevel--;
+            break;
+          case XMLStreamConstants.START_DOCUMENT:
+            break;
+        }
+        if (xmlRecordReady) {
+          currentValue.put(fileName, xmlRecord.toString());
+          return true;
+        }
+      }
+    } catch (XMLStreamException exception) {
+      throw new IllegalArgumentException(exception);
+    }
+    actionsAfterEOF();
+    return false;
+  }
+
+  /**
+   * Method to append start tag information along with attributes if any
+   */
+  private void appendStartTagInformation(String nodeNameStart, StringBuilder xmlRecord) {
+    xmlRecord.append(OPENING_START_TAG_DELIMITER).append(nodeNameStart);
```

Comment:
STAX library does not provide any direct method or way to read start tag with all its attribute, hence doing it manually with string